### ISSUE TITLE
Avoid IIFE definitions in generated JavaScript

### ISSUE THREAD

**Issue by garyb** [purescript/purescript#2206]
Tagging this with discussion, because I'm not sure if doing so is problematic.

An example of the offending code:

``` javascript
var Just = (function () {
    function Just(value0) {
        this.value0 = value0;
    };
    Just.create = function (value0) {
        return new Just(value0);
    };
    return Just;
})();

var Nothing = (function () {
    function Nothing() {
    };
    Nothing.value = new Nothing();
    return Nothing;
})();
```

The IIFE exists so that the definition of `Just` and `Nothing` do not involve statements at the module level (the `create` / `value` property assignments). In theory this makes them _easier_ to erase, since the whole value can be eliminated, but apparently Uglify considers this to be side-effectful. I assume it takes that stance with any IIFE definition, since here only values in the internal scope of the IIFE are referenced.

I actually don't remember exactly why I did introduce the IIFE... it may purely be to avoid the module-level statements, or perhaps codegen can only produce single AST values for declarations. The latter wouldn't be too much of a problem to change, it's just a bit of refactoring.

/cc @AppShipIt @spicydonuts

**Comment by AppShipIt** [comment]
We'll play with it during hacked 🙏

**Comment by spicydonuts** [comment]
The ES6 module discussion is also relevant, since it might not matter whether Uglify removes them if ES6 module "tree shaking" is able to when `exports.Just = ...` is replaced with `export const Just = ...`.

**Comment by AppShipIt** [comment]
> but apparently Uglify considers this to be side-effectful.

Rollup too.

**Comment by garyb** [comment]
When is the next Hacked?

**Comment by paf31** [comment]
This Saturday I think. I plan to work on PSCi, FWIW.

**Comment by michaelficarra** [comment]
Assuming we take #2204 to its logical conclusion, these should be compiled to ES6 classes anyway. Example from #2204:

```js
export class Just {
  constructor(value0) { }
  static create(value0) {
    return new Just(value0);
  }
}
```

**Comment by AppShipIt** [comment]
@michaelficarra @spicydonuts Should psc output equivalent ES5 behind a flag?

**Comment by michaelficarra** [comment]
@AppShipIt I don't think so. If you want to compile modern ECMAScript to old ECMAScript, that's your prerogative, and there's many tools out there which you can include after PureScript in your compilation pipeline.

**Comment by AppShipIt** [comment]
It doesn't affect me personally but I imagine it'd inhibit new users by requiring another dependency.

**Comment by spicydonuts** [comment]
That convenient dev experience seems like the job of something like webpack, possibly pre-configured by pulp

**Comment by AppShipIt** [comment]
https://github.com/purescript/purescript/issues/1206#issuecomment-228581840 may obviate this.

**Comment by Pauan** [comment]
Rollup doesn't have a problem with tree-shaking that code, I don't think IIFEs are a problem anymore.

**Comment by hdgarrood** [comment]
I think they still are: http://rollupjs.org/?version=0.41.4&shareable=JTdCJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmVzJTIyJTJDJTIybW9kdWxlTmFtZSUyMiUzQSUyMm15QnVuZGxlJTIyJTJDJTIyZ2xvYmFscyUyMiUzQSU3QiU3RCU3RCUyQyUyMm1vZHVsZXMlMjIlM0ElNUIlN0IlMjJuYW1lJTIyJTNBJTIybWFpbi5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjIlNUNuJTVDJTIydXNlJTIwc3RyaWN0JTVDJTIyJTNCJTVDbiU1Q24lMkYlMkYlMjBBcyUyMGdlbmVyYXRlZCUyMGJ5JTIwcHNjJTIwMC4xMC41JTVDbnZhciUyMEp1c3QlMjAlM0QlMjAoZnVuY3Rpb24lMjAoKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjBmdW5jdGlvbiUyMEp1c3QodmFsdWUwKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjB0aGlzLnZhbHVlMCUyMCUzRCUyMHZhbHVlMCUzQiU1Q24lMjAlMjAlMjAlMjAlN0QlM0IlNUNuJTIwJTIwJTIwJTIwSnVzdC5jcmVhdGUlMjAlM0QlMjBmdW5jdGlvbiUyMCh2YWx1ZTApJTIwJTdCJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMHJldHVybiUyMG5ldyUyMEp1c3QodmFsdWUwKSUzQiU1Q24lMjAlMjAlMjAlMjAlN0QlM0IlNUNuJTIwJTIwJTIwJTIwcmV0dXJuJTIwSnVzdCUzQiU1Q24lN0QpKCklM0IlNUNuJTVDbiUyRiUyRiUyMEh5cG90aGV0aWNhbCUyMGFsdGVybmF0aXZlJTVDbmZ1bmN0aW9uJTIwSnVzdEFsdCh2YWx1ZTApJTIwJTdCJTVDbiUyMCUyMHRoaXMudmFsdWUwJTIwJTNEJTIwdmFsdWUwJTNCJTVDbiU3RCU1Q25KdXN0QWx0LmNyZWF0ZSUyMCUzRCUyMGZ1bmN0aW9uKHZhbHVlMCklMjAlN0IlNUNuJTIwJTIwcmV0dXJuJTIwbmV3JTIwSnVzdEFsdCh2YWx1ZTApJTNCJTVDbiU3RCU1Q24lNUNuJTJGJTJGJTIwRVMyMDE1JTIwY2xhc3MlNUNuY2xhc3MlMjBKdXN0Q2xhc3MlMjAlN0IlNUNuJTIwJTIwY29uc3RydWN0b3IlMjAodmFsdWUwKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjB0aGlzLnZhbHVlMCUyMCUzRCUyMHZhbHVlMCUzQiU1Q24lMjAlMjAlN0QlNUNuJTIwJTIwJTVDbiUyMCUyMHN0YXRpYyUyMGNyZWF0ZSh2YWx1ZTApJTIwJTdCJTVDbiUyMCUyMCUyMCUyMHJldHVybiUyMG5ldyUyMEp1c3RDbGFzcyh2YWx1ZTApJTNCJTVDbiUyMCUyMCU3RCU1Q24lN0QlNUNuJTVDbiUyRiUyRiUyMFRoaXMlMjBjYW5ub3QlMjBiZSUyMGRldGVybWluZWQlMjB0byUyMGJlJTIwcHVyZSUyMGFuZCUyMGlzJTIwbGVmdCUyMGluJTVDbm5ldyUyMEp1c3QoMCklM0IlNUNuJTVDbiUyRiUyRiUyMFRoaXMlMjBpcyUyMHJlbW92ZWQlNUNubmV3JTIwSnVzdEFsdCgwKSUzQiU1Q24lNUNuJTJGJTJGJTIwVGhpcyUyMGlzJTIwYWxzbyUyMG5vdCUyMHJlbW92ZWQlNUNubmV3JTIwSnVzdENsYXNzKDApJTNCJTVDbiUyMiU3RCU1RCU3RA==

**Comment by Pauan** [comment]
@hdgarrood Right, but I think that's actually a separate issue. [As long as you don't use the data constructor at the top level, then IIFEs are fine.](http://rollupjs.org/?version=0.41.4&shareable=JTdCJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmVzJTIyJTJDJTIybW9kdWxlTmFtZSUyMiUzQSUyMm15QnVuZGxlJTIyJTJDJTIyZ2xvYmFscyUyMiUzQSU3QiU3RCU3RCUyQyUyMm1vZHVsZXMlMjIlM0ElNUIlN0IlMjJuYW1lJTIyJTNBJTIybWFpbi5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjIlNUNuJTVDJTIydXNlJTIwc3RyaWN0JTVDJTIyJTNCJTVDbiU1Q24lMkYlMkYlMjBBcyUyMGdlbmVyYXRlZCUyMGJ5JTIwcHNjJTIwMC4xMC41JTVDbnZhciUyMEp1c3QlMjAlM0QlMjAoZnVuY3Rpb24lMjAoKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjBmdW5jdGlvbiUyMEp1c3QodmFsdWUwKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjB0aGlzLnZhbHVlMCUyMCUzRCUyMHZhbHVlMCUzQiU1Q24lMjAlMjAlMjAlMjAlN0QlM0IlNUNuJTIwJTIwJTIwJTIwSnVzdC5jcmVhdGUlMjAlM0QlMjBmdW5jdGlvbiUyMCh2YWx1ZTApJTIwJTdCJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMHJldHVybiUyMG5ldyUyMEp1c3QodmFsdWUwKSUzQiU1Q24lMjAlMjAlMjAlMjAlN0QlM0IlNUNuJTIwJTIwJTIwJTIwcmV0dXJuJTIwSnVzdCUzQiU1Q24lN0QpKCklM0IlNUNuJTVDbiUyRiUyRiUyMEh5cG90aGV0aWNhbCUyMGFsdGVybmF0aXZlJTVDbmZ1bmN0aW9uJTIwSnVzdEFsdCh2YWx1ZTApJTIwJTdCJTVDbiUyMCUyMHRoaXMudmFsdWUwJTIwJTNEJTIwdmFsdWUwJTNCJTVDbiU3RCU1Q25KdXN0QWx0LmNyZWF0ZSUyMCUzRCUyMGZ1bmN0aW9uKHZhbHVlMCklMjAlN0IlNUNuJTIwJTIwcmV0dXJuJTIwbmV3JTIwSnVzdEFsdCh2YWx1ZTApJTNCJTVDbiU3RCU1Q24lNUNuJTJGJTJGJTIwRVMyMDE1JTIwY2xhc3MlNUNuY2xhc3MlMjBKdXN0Q2xhc3MlMjAlN0IlNUNuJTIwJTIwY29uc3RydWN0b3IlMjAodmFsdWUwKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjB0aGlzLnZhbHVlMCUyMCUzRCUyMHZhbHVlMCUzQiU1Q24lMjAlMjAlN0QlNUNuJTIwJTIwJTVDbiUyMCUyMHN0YXRpYyUyMGNyZWF0ZSh2YWx1ZTApJTIwJTdCJTVDbiUyMCUyMCUyMCUyMHJldHVybiUyMG5ldyUyMEp1c3RDbGFzcyh2YWx1ZTApJTNCJTVDbiUyMCUyMCU3RCU1Q24lN0QlNUNuJTVDbmZ1bmN0aW9uJTIwZm9vKCklMjAlN0IlNUNuJTIwJTIwcmV0dXJuJTIwbmV3JTIwSnVzdCgwKSUzQiU1Q24lN0QlNUNuJTIyJTdEJTVEJTdE)

The reason why it's not tree shaking is because [my pull request](https://github.com/rollup/rollup/pull/1017) adds in very limited detection for ES5 classes, and the IIFEs mess up that detection.

In addition, my pull request only applies to `new Foo(...)`, [so it doesn't help at all with `Just.create(1)`](http://rollupjs.org/?version=0.41.4&shareable=JTdCJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmVzJTIyJTJDJTIybW9kdWxlTmFtZSUyMiUzQSUyMm15QnVuZGxlJTIyJTJDJTIyZ2xvYmFscyUyMiUzQSU3QiU3RCU3RCUyQyUyMm1vZHVsZXMlMjIlM0ElNUIlN0IlMjJuYW1lJTIyJTNBJTIybWFpbi5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjIlNUNuJTVDJTIydXNlJTIwc3RyaWN0JTVDJTIyJTNCJTVDbiU1Q24lMkYlMkYlMjBBcyUyMGdlbmVyYXRlZCUyMGJ5JTIwcHNjJTIwMC4xMC41JTVDbnZhciUyMEp1c3QlMjAlM0QlMjAoZnVuY3Rpb24lMjAoKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjBmdW5jdGlvbiUyMEp1c3QodmFsdWUwKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjB0aGlzLnZhbHVlMCUyMCUzRCUyMHZhbHVlMCUzQiU1Q24lMjAlMjAlMjAlMjAlN0QlM0IlNUNuJTIwJTIwJTIwJTIwSnVzdC5jcmVhdGUlMjAlM0QlMjBmdW5jdGlvbiUyMCh2YWx1ZTApJTIwJTdCJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMHJldHVybiUyMG5ldyUyMEp1c3QodmFsdWUwKSUzQiU1Q24lMjAlMjAlMjAlMjAlN0QlM0IlNUNuJTIwJTIwJTIwJTIwcmV0dXJuJTIwSnVzdCUzQiU1Q24lN0QpKCklM0IlNUNuJTVDbiUyRiUyRiUyMEh5cG90aGV0aWNhbCUyMGFsdGVybmF0aXZlJTVDbmZ1bmN0aW9uJTIwSnVzdEFsdCh2YWx1ZTApJTIwJTdCJTVDbiUyMCUyMHRoaXMudmFsdWUwJTIwJTNEJTIwdmFsdWUwJTNCJTVDbiU3RCU1Q25KdXN0QWx0LmNyZWF0ZSUyMCUzRCUyMGZ1bmN0aW9uKHZhbHVlMCklMjAlN0IlNUNuJTIwJTIwcmV0dXJuJTIwbmV3JTIwSnVzdEFsdCh2YWx1ZTApJTNCJTVDbiU3RCU1Q24lNUNuJTJGJTJGJTIwRVMyMDE1JTIwY2xhc3MlNUNuY2xhc3MlMjBKdXN0Q2xhc3MlMjAlN0IlNUNuJTIwJTIwY29uc3RydWN0b3IlMjAodmFsdWUwKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjB0aGlzLnZhbHVlMCUyMCUzRCUyMHZhbHVlMCUzQiU1Q24lMjAlMjAlN0QlNUNuJTIwJTIwJTVDbiUyMCUyMHN0YXRpYyUyMGNyZWF0ZSh2YWx1ZTApJTIwJTdCJTVDbiUyMCUyMCUyMCUyMHJldHVybiUyMG5ldyUyMEp1c3RDbGFzcyh2YWx1ZTApJTNCJTVDbiUyMCUyMCU3RCU1Q24lN0QlNUNuJTVDbiUyRiUyRiUyMFRoaXMlMjBjYW5ub3QlMjBiZSUyMGRldGVybWluZWQlMjB0byUyMGJlJTIwcHVyZSUyMGFuZCUyMGlzJTIwbGVmdCUyMGluJTVDbm5ldyUyMEp1c3QoMCklM0IlNUNuJTVDbkp1c3QuY3JlYXRlKDEpJTNCJTVDbiU1Q24lMkYlMkYlMjBUaGlzJTIwaXMlMjByZW1vdmVkJTVDbm5ldyUyMEp1c3RBbHQoMCklM0IlNUNuJTVDbkp1c3RBbHQuY3JlYXRlKDEpJTNCJTVDbiU1Q24lMkYlMkYlMjBUaGlzJTIwaXMlMjBhbHNvJTIwbm90JTIwcmVtb3ZlZCU1Q25uZXclMjBKdXN0Q2xhc3MoMCklM0IlNUNuJTVDbkp1c3RDbGFzcy5jcmVhdGUoMSklM0IlNUNuJTIyJTdEJTVEJTdE), even if you remove the IIFE.

**Comment by hdgarrood** [comment]
I think it's being removed in that example because Rollup can determine that it's unreachable, since your `foo()` is never called. If you put in a call to `foo()`, it still makes it into the bundle.

I think having to avoid using data constructors at the top level would be an acceptable caveat for a tool that is going to process PureScript code, as any given program can be expected to contain lots of unused top level declarations which mention data constructors, and these declarations should be eliminated by DCE.

**Comment by hdgarrood** [comment]
To clarify, I think the real issue is that DCE tools are not able to determine that both the definitions of data constructors are pure and uses of those data constructors are pure in the generated JS. We need DCE tools to be able to determine that both of these properties hold in order to perform good quality DCE. One way of addressing this would be to get rid of the IIFEs in the generated JS, but it might not be the only one. It is the only one I'm aware of, though.

**Comment by garyb** [comment]
Since the dctor IIFEs serve no real purpose and this would be a pretty trivial change to make I'm not sure why we're still talking about it 😆.

They're a hangover from the days where PS modules were declared as one giant object rather than having an export list at the end.

**Comment by hdgarrood** [comment]
We don't even use the `.create()` methods, do we?

**Comment by garyb** [comment]
Yes, as we don't require constructors to be fully applied - the `create` is the curried version. We could probably drop them for one-argument constructors though.

**Comment by Pauan** [comment]
@hdgarrood You're right about `foo`, so ultimately it comes down to the limited detection in my pull request.

@garyb Great, I think it would be a good idea to remove the IIFEs. That's a good 80% solution.

**Comment by paf31** [comment]
Reopening, since the PR was reverted. We should consider merging this back in once psc-bundle is updated for 1.0 (i.e. once Rollup is the supported code stripping tool).

**Comment by hdgarrood** [comment]
:+1: it would be nice to update psc-bundle so that it can do DCE properly on the no-iife versions of data constructors, but given that it's more tricky than I thought I'm tempted to just leave data constructor codegen with the IIFEs until after we have removed DCE from psc-bundle &mdash; I thought I had done enough testing to rule out the kind of bug we've just seen but obviously not.

**Comment by kevinbarabash** [comment]
It seems like there's many more top level data constructors than IIFEs so figuring out how to DCE the data constructors should provide a bigger win.
